src/Sarif.Viewer.VisualStudio/Models/SarifErrorListItem.cs

```diff
@@ -597,112 +602,100 @@ internal void RemapFilePath(string originalPath, string remappedPath)
         /// <summary>
         /// Attaches to the document using the specified properties, which are also cached.
         /// </summary>
-        internal void AttachToDocument(string documentName, long docCookie, IVsWindowFrame windowFrame)
+        internal bool TryAttachToDocument(string documentName, long docCookie, IVsWindowFrame windowFrame)
         {
             ThreadHelper.ThrowIfNotOnUIThread();
+
+            if (_documentCookie.HasValue || string.Compare(documentName, this.FileName, StringComparison.OrdinalIgnoreCase) != 0)
+            {
+                return false;
+            }
+
             // Cache the document info so we can detach and reattach later.
             _documentName = documentName;
             _documentCookie = docCookie;
             _windowFrame = windowFrame;
 
             AttachToDocument();
+
+            return true;
         }
 
         /// <summary>
         /// Attaches to the document using cached properties.
         /// </summary>
-        internal void AttachToDocument()
+        private void AttachToDocument()
         {
             ThreadHelper.ThrowIfNotOnUIThread();
-            LineMarker?.AttachToDocument(_documentName, _documentCookie, _windowFrame);
 
-            foreach (LocationModel location in Locations)
-            {
-                location.LineMarker?.AttachToDocument(_documentName, _documentCookie, _windowFrame);
+            IComponentModel componentModel = (IComponentModel)AsyncPackage.GetGlobalService(typeof(SComponentModel));
+            if (componentModel == null)
+            { 
+                return;
             }
 
-            foreach (LocationModel location in RelatedLocations)
-            {
-                location.LineMarker?.AttachToDocument(_documentName, _documentCookie, _windowFrame);
-            }
+            ISarifLocationProviderFactory sarifLocationProviderFactory = componentModel.GetService<ISarifLocationProviderFactory>();
 
-            foreach (CallTree callTree in CallTrees)
+            // Get a SimpleTagger over the buffer to color
+            IVsTextView vsTextView = SdkUIUtilities.GetTextViewFromFrame(_windowFrame);
+            IWpfTextView wpfTextView = SdkUIUtilities.GetWpfTextView(vsTextView);
+            SimpleTagger<TextMarkerTag> tagger = sarifLocationProviderFactory.GetTextMarkerTagger(wpfTextView.TextBuffer);
+
+            using (tagger.Update())
             {
-                Stack<CallTreeNode> nodesToProcess = new Stack<CallTreeNode>();
+                LineMarker?.TryAttachToDocument(_documentName, _windowFrame);
 
-                foreach (CallTreeNode topLevelNode in callTree.TopLevelNodes)
+                foreach (LocationModel location in Locations)
                 {
-                    nodesToProcess.Push(topLevelNode);
+                    location.LineMarker?.TryAttachToDocument(_documentName, _windowFrame);
```

Comment:
If try works should this not break?

Reply:
because you can have multiple locations for the same document. So you must continue.

---
In reply to: [479670554](https://github.com/microsoft/sarif-visualstudio-extension/pull/186#discussion_r479670554) [](ancestors = 479670554)